src/module.c

```diff
@@ -8359,8 +8367,11 @@ void moduleHandleBlockedClients(void) {
         /* Update stats now that we've finished the blocking operation.
          * This needs to be out of the reply callback above given that a
          * module might not define any callback and still do blocking ops.
+         *
+         * If the module is blocked on keys updateStatsOnUnblock will be
+         * called from moduleUnblockClientOnKey
          */
-        if (c && !clientHasModuleAuthInProgress(c)) {
+        if (c && !clientHasModuleAuthInProgress(c) && !bc->blocked_on_keys) {
             int had_errors = c->deferred_reply_errors ? !!listLength(c->deferred_reply_errors) :
                 (server.stat_total_error_replies != prev_error_replies);
             updateStatsOnUnblock(c, bc->background_duration, reply_us, had_errors);
```

Comment:
does these need to be reverted too? we no longer use `REDISMODULE_CTX_THREAD_SAFE` in `moduleBlockedClientTimedOut`.

```suggestion
        if (c && !clientHasModuleAuthInProgress(c) && !bc->blocked_on_keys) {
            updateStatsOnUnblock(c, bc->background_duration, reply_us, server.stat_total_error_replies != prev_error_replies);
```